cips/cip-shwap_protocol.md

```diff
@@ -0,0 +1,407 @@
+capitals, as shown here.
+
+Commonly used terms in this document are described below.
+
+_**Shwap**_: The protocol described by this document. Shwap is a
+portmanteau name of words share and swap.
+
+_**[Share][shares]**_: The core data structure of DataSquare **"swapped"** between peers.
+
+_**[DataSquare][square]**_: The DA square format used by Celestia DA network.
+
+_**[DAH][dah]**_: The Data Availability Header with Row and Column commitments.
+
+_**[Namespace][ns]**_: The namespace grouping sets of shares.
+
+_**Peer**_: An entity that can participate in a Shwap protocol. There are three types of peers:
+client, server, and node.
+
+_**Client**_: The Peer that requests content by share identifiers over Shwap.
+
+_**Server**_: The Peer that responds with shares over Shwap.
+
+_**Node**_: The peer that is both the client and the server.
+
+_**Proof**_: A Merkle inclusion proof of the data in the DataSquare.
+
+### Message Framework
+
+This section defines Shwap's messaging framework. Every group of shares that needs to be exchanged over the network
+MUST define its [share identifier](#share-identifiers) and [share container](#share-containers) and follow their
+described rules. Every identifier and container MUST define its serialization format, which MAY NOT be consistent with
+other identifiers and containers.
+
+
+#### Share Containers
+
+Share containers encapsulate a set of data shares with proof. Share containers are identified by [share identifiers](#share-identifiers).
+
+Containers SHOULD contain shares within a single DataSquare and MAY NOT be adjacent. Containers MUST have a [DAH][dah]
+inclusion proof field defined.
+
+##### Serialization
+
+Share containers are RECOMMENDED to use protobuf (proto3) encoding, and other formats MAY be used for serialization. A 
+container MAY define multiple serialization formats.
+
+#### Share Identifiers
+
+Share identifiers identify [share containers](#share-containers). Identifiers are not collision-resistant and there MAY 
+be different identifiers referencing the same container.
+
+Identifiers MAY embed each other to narrow down the scope of needed shares. For example, [SampleID](#sampleid) embeds
+[RowID](#rowid) as every sample lay on a particular row.
+
+##### Serialization
+
+Share identifiers SHOULD be serialized by concatenating big-endian representations of fields in the order defined by 
+their respective formatting section. Serialized identifiers SHOULD have constant and predetermined lengths in bytes.
+
+#### Versioning
+
+If a defined share container or identifier requires an incompatible change, the new message type MAY be introduced
+suffixed with a new major version starting from v1. E.g., if the Row message needs a revision, RowV1 is created.
+
+### Messages
+
+This section defines all the supported Shwap messages, including share identifiers and containers. All the new
+future messages should be described in the section.
```

Comment:
[question] in which section?

Reply:
This

Reply:
ahh so `the` is just a typo and should be `this`